File: Objective-C/advanced/APIDemo/APIDemo/AdMobAdTargetingTableViewController.m
//
//  Copyright (C) 2015 Google, Inc.
//
//  AdMobAdTargetingTableViewController.m
//  APIDemo
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "AdMobAdTargetingTableViewController.h"
#import <GoogleMobileAds/GoogleMobileAds.h>

#import "Constants.h"

/// The constants for table cell identifiers.
static NSString *const kAgeRestrictedCellIdentifier = @"ageRestrictedCell";
static NSString *const kAgeRestrictedPickerCellIdentifier = @"ageRestrictedPickerCell";

/// AdMob - Ad Targeting
/// Demonstrates AdMob ad targeting.
@interface AdMobAdTargetingTableViewController () <UIPickerViewDelegate,
                                                   UIPickerViewDataSource,
                                                   GADBannerViewDelegate>

/// The age-restricted treatment label.
@property(nonatomic, weak) IBOutlet UILabel *ageRestrictedLabel;

/// The age-restricted treatment picker.
@property(nonatomic, weak) IBOutlet UIPickerView *ageRestrictedPicker;

/// The banner view.
@property(nonatomic, weak) IBOutlet GADBannerView *bannerView;

/// Loads an ad based on age-restricted treatment.
- (IBAction)loadTargetedAd:(id)sender;

/// The age-restricted treatment options.
@property(nonatomic, copy) NSArray<NSString *> *ageRestrictedOptions;

@end

@implementation AdMobAdTargetingTableViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  // Collapse table view footer view, display default section color for table view background.
  // Styling for larger screen sizes.
  self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
  UIColor *backgroundColor =
      [UIColor colorWithRed:247 / 255.0 green:247 / 255.0 blue:247 / 255.0 alpha:1.0];
  self.tableView.backgroundColor = backgroundColor;

  self.ageRestrictedOptions = @[ @"Child", @"Teen", @"Unspecified" ];
  self.ageRestrictedPicker.delegate = self;
  self.ageRestrictedPicker.dataSource = self;
  [self.ageRestrictedPicker selectRow:0 inComponent:0 animated:NO];
  self.ageRestrictedLabel.text = self.ageRestrictedOptions[0];
}

#pragma mark - UITableViewDataSource

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [self tableView:tableView cellForRowAtIndexPath:indexPath];
  NSString *cellIdentifier = cell.reuseIdentifier;

  if ([cellIdentifier isEqual:kAgeRestrictedPickerCellIdentifier] &&
      self.ageRestrictedPicker.hidden) {
    return 0;
  }
  return [super tableView:tableView heightForRowAtIndexPath:indexPath];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
  NSString *cellIdentifier = cell.reuseIdentifier;
  UIView *currentPickerView;

  if ([cellIdentifier isEqual:kAgeRestrictedCellIdentifier]) {
    currentPickerView = self.ageRestrictedPicker;
  }

  BOOL isPickerHidden = currentPickerView.hidden;
  [self hideAllPickers];
  currentPickerView.hidden = !isPickerHidden;
  [tableView reloadData];
  [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)tableView:(UITableView *)tableView
    willDisplayHeaderView:(UIView *)view
               forSection:(NSInteger)section {
  view.tintColor = [UIColor clearColor];
}

#pragma mark - UIPickerViewDataSource

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  NSInteger numberOfRows = 0;
  if (pickerView == self.ageRestrictedPicker) {
    numberOfRows = self.ageRestrictedOptions.count;
  }
  return numberOfRows;
}

#pragma mark - UIPickerViewDelegate

- (NSString *)pickerView:(UIPickerView *)pickerView
             titleForRow:(NSInteger)row
            forComponent:(NSInteger)component {
  NSString *titleForRow = @"";
  if (pickerView == self.ageRestrictedPicker) {
    titleForRow = self.ageRestrictedOptions[row];
  }
  return titleForRow;
}

- (void)pickerView:(UIPickerView *)pickerView
      didSelectRow:(NSInteger)row
       inComponent:(NSInteger)component {
  if (pickerView == self.ageRestrictedPicker) {
    self.ageRestrictedLabel.text = self.ageRestrictedOptions[row];
  }
}

#pragma mark - Actions

- (IBAction)loadTargetedAd:(id)sender {
  self.bannerView.adUnitID = AdUnitIDBanner;
  self.bannerView.rootViewController = self;
  self.bannerView.delegate = self;

  GADRequest *request = [GADRequest request];
  if ([self.ageRestrictedLabel.text isEqual:@"Child"]) {
    GADMobileAds.sharedInstance.requestConfiguration.ageRestrictedTreatment =
        GADAgeRestrictedTreatmentChild;
  } else if ([self.ageRestrictedLabel.text isEqual:@"Teen"]) {
    GADMobileAds.sharedInstance.requestConfiguration.ageRestrictedTreatment =
        GADAgeRestrictedTreatmentTeen;
  } else {
    GADMobileAds.sharedInstance.requestConfiguration.ageRestrictedTreatment =
        GADAgeRestrictedTreatmentUnspecified;
  }

  [self.bannerView loadRequest:request];
}

- (void)hideAllPickers {
  self.ageRestrictedPicker.hidden = YES;
}

#pragma mark - GADBannerViewDelegate

- (void)adViewDidReceiveAd:(GADBannerView *)bannerView {
  NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)adView:(GADBannerView *)bannerView didFailToReceiveAdWithError:(NSError *)error {
  NSLog(@"%s: %@", __PRETTY_FUNCTION__, error.localizedDescription);
}

@end

File: Swift/advanced/APIDemo/APIDemo/AdMobAdTargetingTableViewController.swift
//
//  Copyright (C) 2016 Google, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import GoogleMobileAds
import UIKit

/// The constants for AdMobAdTargeting table cell identifiers.
struct AdMobAdTargetingTableCellIdentifiers {
  static let ageRestrictedCell = "ageRestrictedCell"
  static let ageRestrictedPickerCell = "ageRestrictedPickerCell"
}

/// AdMob - Ad Targeting
/// Demonstrates AdMob ad targeting.
class AdMobAdTargetingTableViewController: UITableViewController, UIPickerViewDataSource,
  UIPickerViewDelegate
{

  /// The age-restricted treatment label.
  @IBOutlet weak var ageRestrictedLabel: UILabel!

  /// The age-restricted treatment picker.
  @IBOutlet weak var ageRestrictedPicker: UIPickerView!

  /// The banner view.
  @IBOutlet weak var bannerView: BannerView!

  /// The age-restricted treatment options.
  var ageRestrictedOptions: [String]!

  override func viewDidLoad() {
    super.viewDidLoad()
    tableView.tableFooterView = UIView(frame: CGRect.zero)

    // GADBannerView setup.
    bannerView.adUnitID = Constants.adMobAdUnitID
    bannerView.rootViewController = self
    bannerView.delegate = self

    // Age-restricted setup.
    ageRestrictedOptions = ["Child", "Teen", "Unspecified"]
    ageRestrictedPicker.delegate = self
    ageRestrictedPicker.dataSource = self
    ageRestrictedPicker.selectRow(
      0, inComponent: 0,
      animated: false)
    ageRestrictedLabel.text = ageRestrictedOptions[0]
  }

  // MARK: - UITableViewDelegate

  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    let cell = tableView.cellForRow(at: indexPath)
    var currentPicker: UIView?
    if let cellIdentifier = cell?.reuseIdentifier {
      switch cellIdentifier {
      case AdMobAdTargetingTableCellIdentifiers.ageRestrictedCell:
        currentPicker = ageRestrictedPicker
      default:
        break
      }
    }
    if let isPickerHidden = currentPicker?.isHidden {
      hideAllPickers()
      currentPicker?.isHidden = !isPickerHidden
      tableView.reloadData()
      tableView.deselectRow(at: indexPath, animated: true)
    }
  }

  override func tableView(
    _ tableView: UITableView, willDisplayHeaderView view: UIView,
    forSection section: Int
  ) {
    view.tintColor = UIColor.clear
  }

  // MARK: - UITableViewDataSource

  override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath)
    -> CGFloat
  {
    let cell = self.tableView(tableView, cellForRowAt: indexPath)
    if let cellIdentifier = cell.reuseIdentifier {
      if cellIdentifier == AdMobAdTargetingTableCellIdentifiers.ageRestrictedPickerCell
        && ageRestrictedPicker.isHidden
      {
        return 0
      }
    }
    return super.tableView(tableView, heightForRowAt: indexPath)
  }

  // MARK: - UIPickerViewDelegate

  func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int)
    -> String?
  {
    var rowTitle = ""
    switch pickerView {
    case ageRestrictedPicker:
      rowTitle = ageRestrictedOptions[row]
    default:
      rowTitle = ""
    }
    return rowTitle
  }

  func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
    switch pickerView {
    case ageRestrictedPicker:
      ageRestrictedLabel.text = ageRestrictedOptions[row]
    default:
      break
    }
  }

  // MARK: - UIPickerViewDataSource

  func numberOfComponents(in pickerView: UIPickerView) -> Int {
    return 1
  }

  func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
    var numOfRows = 0
    switch pickerView {
    case ageRestrictedPicker:
      numOfRows = ageRestrictedOptions.count
    default:
      numOfRows = 0
    }
    return numOfRows
  }

  // MARK: - Actions

  /// Loads an ad based on age-restricted treatment.
  @IBAction func loadTargetedAd(_ sender: AnyObject) {
    let request = Request()
    switch ageRestrictedLabel.text {
    case "Child":
      MobileAds.shared.requestConfiguration.ageRestrictedTreatment = .child
    case "Teen":
      MobileAds.shared.requestConfiguration.ageRestrictedTreatment = .teen
    default:
      MobileAds.shared.requestConfiguration.ageRestrictedTreatment = .unspecified
    }
    bannerView.load(request)
  }

  fileprivate func hideAllPickers() {
    ageRestrictedPicker.isHidden = true
  }

}

extension AdMobAdTargetingTableViewController: BannerViewDelegate {

  func bannerViewDidReceiveAd(_ bannerView: BannerView) {
    print("\(#function)")
  }

  func bannerView(_ bannerView: BannerView, didFailToReceiveAdWithError error: Error) {
    print("\(#function): \(error.localizedDescription)")
  }
}
